subsys/retention/retention.c

```diff
@@ -0,0 +1,400 @@
+		pos = 0;
+
+		while (pos < config->prefix_len) {
+			uint8_t read_size = MIN((config->prefix_len - pos), sizeof(buffer));
+
+			rc = retained_mem_read(config->parent, (config->offset + pos), buffer,
+					       read_size);
+
+			if (rc < 0) {
+				goto finish;
+			}
+
+			if (memcmp(&config->prefix[pos], buffer, read_size) != 0) {
+				/* If the magic header does not match, do not check the rest of
+				 * the validity of the data, assume it is invalid
+				 */
+				data->header_written = false;
+				rc = 0;
+				goto finish;
+			}
+
+			pos += read_size;
+		}
+
+		/* Header already exists so no need to re-write it again */
+		data->header_written = true;
+	}
+
+	if (config->checksum != 0) {
+		/* Check the checksum validity, for this all the data must be read out */
+		uint32_t checksum = 0;
+		uint32_t expected_checksum = 0;
+		ssize_t data_size = config->size - config->checksum;
+
+		rc = retention_checksum(dev, &checksum);
+
+		if (rc < 0) {
+			goto finish;
+		}
+
+		if (config->checksum == CHECKSUM_TYPE_CRC8) {
+			uint8_t read_checksum;
+
+			rc = retained_mem_read(config->parent, (config->offset + data_size),
+					       (void *)&read_checksum, sizeof(read_checksum));
+			expected_checksum = (uint32_t)read_checksum;
+		} else if (config->checksum == CHECKSUM_TYPE_CRC16) {
+			uint16_t read_checksum;
+
+			rc = retained_mem_read(config->parent, (config->offset + data_size),
+					       (void *)&read_checksum, sizeof(read_checksum));
+			expected_checksum = (uint32_t)read_checksum;
+		} else if (config->checksum == CHECKSUM_TYPE_CRC32) {
+			rc = retained_mem_read(config->parent, (config->offset + data_size),
+					       (void *)&expected_checksum,
+					       sizeof(expected_checksum));
+		}
+
+		if (rc < 0) {
+			goto finish;
+		}
+
+		if (checksum != expected_checksum) {
+			goto finish;
+		}
+	}
+
+	/* At this point, checks have passed (if enabled), mark data as being valid */
+	rc = DATA_VALID_VALUE;
+
+finish:
+	retention_lock_release(dev);
+
+	return rc;
+}
+
+int retention_read(const struct device *dev, off_t offset, uint8_t *buffer, size_t size)
+{
+	const struct retention_config *config = dev->config;
+	int rc;
+
+	if (offset < 0 || ((size_t)offset + size) > (config->size - config->reserved_size)) {
+		/* Disallow reading past the virtual data size or before it */
+		return -EINVAL;
+	}
+
+	retention_lock_take(dev);
+
+	rc = retained_mem_read(config->parent, (config->offset + config->prefix_len +
+			       (size_t)offset), buffer, size);
+
+	retention_lock_release(dev);
+
+	return rc;
+}
+
+int retention_write(const struct device *dev, off_t offset, const uint8_t *buffer, size_t size)
+{
+	const struct retention_config *config = dev->config;
+	struct retention_data *data = dev->data;
+	int rc;
+
+	retention_lock_take(dev);
+
+	if (offset < 0 || ((size_t)offset + size) > (config->size - config->reserved_size)) {
+		/* Disallow reading past the virtual data size or before it */
```

Comment:
Need to change in comment `writing` not `reading`

Reply:
Fixed